keystone_plugin/project.py

```diff
@@ -21,60 +20,52 @@
                                      with_neutron_client,
-                                     get_resource_id,
+                                     get_openstack_id,
                                      use_external_resource,
                                      delete_resource_and_runtime_properties,
+                                     add_list_to_runtime_properties,
                                      validate_resource,
-                                     COMMON_RUNTIME_PROPERTIES_KEYS,
-                                     OPENSTACK_ID_PROPERTY,
-                                     OPENSTACK_TYPE_PROPERTY,
-                                     OPENSTACK_NAME_PROPERTY)
+                                     create_object_dict,
+                                     set_openstack_runtime_properties,
+                                     COMMON_RUNTIME_PROPERTIES_KEYS)
 
 
 PROJECT_OPENSTACK_TYPE = 'project'
 
-TENANT_QUOTA_TYPE = 'quota'
+PROJECT_QUOTA_TYPE = 'quota'
 
 RUNTIME_PROPERTIES_KEYS = COMMON_RUNTIME_PROPERTIES_KEYS
 
+NOVA = 'nova'
+CINDER = 'cinder'
+NEUTRON = 'neutron'
+
+QUOTA = 'quota'
+USERS = 'users'
+ROLES = 'roles'
+
 
 @operation
 @with_keystone_client
-def create(keystone_client, **kwargs):
+def create(keystone_client, args, **kwargs):
     if use_external_resource(ctx, keystone_client, PROJECT_OPENSTACK_TYPE):
         return
 
-    project_dict = {
-        'name': get_resource_id(ctx, PROJECT_OPENSTACK_TYPE),
-        'domain': 'default'
-    }
+    project_dict = create_object_dict(ctx, PROJECT_OPENSTACK_TYPE, args)
+    project_dict['domain'] = 'default'
 
-    project_dict.update(ctx.node.properties['project'])
     project = keystone_client.projects.create(**project_dict)
-
-    ctx.instance.runtime_properties[OPENSTACK_ID_PROPERTY] = project.id
-    ctx.instance.runtime_properties[OPENSTACK_TYPE_PROPERTY] = \
-        PROJECT_OPENSTACK_TYPE
-    ctx.instance.runtime_properties[OPENSTACK_NAME_PROPERTY] = project.name
+    set_openstack_runtime_properties(ctx, project, PROJECT_OPENSTACK_TYPE)
 
 
 @operation
-@with_keystone_client
-@with_nova_client
-@with_cinder_client
-@with_neutron_client
-def start(keystone_client, nova_client, cinder_client, neutron_client,
-          **kwargs):
-    project_id = ctx.instance.runtime_properties[OPENSTACK_ID_PROPERTY]
-    users = ctx.node.properties['users']
-    validate_users(users, keystone_client)
-
-    assign_users(project_id, users, keystone_client)
-
-    quota = ctx.node.properties[TENANT_QUOTA_TYPE]
-    update_quota(project_id, quota, nova_client, 'nova')
-    update_quota(project_id, quota, neutron_client, 'neutron')
-    update_quota(project_id, quota, cinder_client, 'cinder')
+def start(quota_dict, **kwargs):
+    users = ctx.node.properties[USERS]
+    validate_users(users, **kwargs)
+    assign_users(users, **kwargs)
+    quota = ctx.node.properties[PROJECT_QUOTA_TYPE]
+    quota.update(quota_dict)
+    update_project_quota(quota=quota, **kwargs)
```

Comment:
I am wondering, what happens if `update_project_quota` fails but assign_users already succeeds.
There is no `NonRecoverableError` raised, so this operation will be retried. Then we will try to create the user again. Then this will fail. At the very least we should raise a `NonRecoverableError`. Ideally, we would do some kind of rollback. Although really, that should be something that Cloudify does.

Reply:
Good point